### PR TITLE
[NVIDIA XLA:GPU] Fused MHA rewriter

### PR DESCRIPTION
**Pull request by AyanmoI**
cuDNN currently supports and the rewriter matches the following patterns for Multi-headed attention:
1. BMM1 - BMM2
2. BMM1 - Scale - Bias - Mask - Softmax - BMM2 
3. BMM1 - Scale - Bias - Mask - Softmax - Dropout - BMM2
4. BMM1 - Scale - Mask - Softmax - BMM2
5. BMM1 - Scale - Mask - Softmax - Dropout - BMM2 
6. BMM1 - Softmax - Dropout - BMM2 

This PR adds rewriter support for the first pattern. 
The following pattern is matched here:
`Output = BMM2(BMM1(Q,K), V)`
The other infrastructure piece is in https://github.com/tensorflow/tensorflow/pull/59154

### CODE REVIEW

**Comment by jprabhas** [comment]
Why not add an HLO based test for the rewriter ?

**Comment by cheshire** [comment]
As far as I understand this is not meant for review yet, and more like an FYI, right?

**Comment by AyanmoI** [comment]
> As far as I understand this is not meant for review yet, and more like an FYI, right?

No this is actually up for review. This is the starter PR than can go in independently. We will subsequently add more pieces.

**Comment by AyanmoI** [comment]
@cheshire I am not sure why the tests are failing with
```2023-01-30 00:12:47.838404: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:266] failed call to cuInit: UNKNOWN ERROR (34)
2023-01-30 00:12:47.838996: F tensorflow/tsl/platform/statusor.cc:33] Attempting to fetch value instead of handling error NOT_FOUND: no CUDA devices found```
Is it running on machines that it is not supposed to? 

Also, this PR is ready for review. Not sure if you have already started looking at this though.

**Comment by cheshire** [comment]
> Attempting to fetch value instead of handling error NOT_FOUND: no CUDA devices found

You probably need to add gpu_plugin as a dependency.

File: tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc [hunk 1]
```diff
@@ -51,6 +51,19 @@ const absl::string_view kCudnnConvBiasActivationForwardCallTarget =
 const absl::string_view kCudnnConvForwardCallTarget = "__cudnn$convForward";
 const absl::string_view kTriangularSolveCallTarget = "__cublas$triangularSolve";
 
+// fMHA call targets.
+const absl::string_view kCudnnfMHADefaultCallTarget = "__cudnn$fhmaDefault";
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc:55]: Would it make more sense to use a unified call target and push those inside backend_config instead?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+  TF_ASSIGN_OR_RETURN(
+      std::vector<int64_t> rhs_non_contracting_dim_nums_bmm2,
+      GetNonContractingDims(bmm_2->operand(1)->shape(),
+                            dot_dims_bmm2.rhs_batch_dimensions(),
+                            dot_dims_bmm2.rhs_contracting_dimensions()));
+
+  std::vector<int64_t> rhs_non_contracting_dims_bmm2 =
+      GetDimensionVector(bmm_2->operand(1)->shape().dimensions(),
+                         rhs_non_contracting_dim_nums_bmm2);
+  // The non contracting dimension for BMM2 needs to be 64 for the input matrix.
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:226]: What's the significance of 64? Is that how fMHA looks in practice in LLMs?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  ROOT custom-call.1 = bf16[20,40,64]{2,1,0} custom-call(custom-call, Arg_2.3), custom_call_target="__cublas$gemm", backend_config="{\"alpha_real\":1,\"alpha_imag\":0,\"beta\":0,\"dot_dimension_numbers\":{\"lhs_contracting_dimensions\":[\"2\"],\"rhs_contracting_dimensions\":[\"1\"],\"lhs_batch_dimensions\":[\"0\"],\"rhs_batch_dimensions\":[\"0\"]},\"precision_config\":{\"operand_precision\":[\"DEFAULT\",\"DEFAULT\"]},\"epilogue\":\"DEFAULT\"}"
+}
+
+
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+  CudnnFusedMHARewriter fusedMhaRewriter{GetCudaComputeCapability()};
+  TF_ASSERT_OK_AND_ASSIGN(bool result, RunHloPass(&fusedMhaRewriter, m.get()));
+  EXPECT_FALSE(result);
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:176]: Should we also check for correctness and run the resulting HLO?

**Review by cheshire — CHANGES_REQUESTED** [review]
Don't you need extra plumbing to then dispatch resulting fusions?

File: tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc:55]: Since `call_target` is an existing attribute, I felt it made sense to have separate call targets rather and adding another field to backend_config. Moreover, this is the convention followed by convolution as well.  Do you strongly suggest changing this?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc (hunk 3 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:176]: This is simple negative test that makes sure that the rewriter doesn't trigger a rewrite when the conditions are not end. I'll add functional correctness test with the upcoming PR that can run e2e. Would you prefer to add a correctness check here to make fmha lowering has not been triggered?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 2 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:226]: 64 relates to the L1 cache and/or shared memory constraint in the kernel implemented by the cuDNN team. I'll find a more comprehensive answer for this. FMHA in current LLMs (certainly T5x) do look like this. So we address the immediate goal of improving perf of current crop of LLMs.

**Comment by AyanmoI** [comment]
Closing this since https://github.com/tensorflow/tensorflow/pull/59832 has been opened that combines this and https://github.com/tensorflow/tensorflow/pull/59154
